bndtools.m2e/src/bndtools/m2e/BndConfigurator.java

```diff
@@ -163,50 +174,87 @@ public AbstractBuildParticipant getBuildParticipant(final IMavenProjectFacade pr
 				// during maven builder will throw lifecycle
 				// errors
 
-				Job job = new WorkspaceJob("Executing " + project.getName() + " jar:jar goal") {
+				Job job = new WorkspaceJob(
+					"Executing " + project.getName() + " jar:" + (isTest ? "test-" : "") + "jar goal") {
 					@Override
 					public IStatus runInWorkspace(IProgressMonitor monitor) throws CoreException {
-						SubMonitor progress = SubMonitor.convert(monitor, 3);
-						execJarMojo(projectFacade, progress.newChild(1, SubMonitor.SUPPRESS_NONE));
-
-						IPath targetDirPath = Path.fromOSString(targetDirectory);
-						IPath projectPath = project.getLocation();
-						IPath relativeTargetDirPath = targetDirPath.makeRelativeTo(projectPath);
-						IFolder targetDir = project.getFolder(relativeTargetDirPath);
-
-						boolean needsRefresh = deltaFilePaths.stream()
-							.filter(deltaPath -> !targetDir.getFullPath()
-								.isPrefixOf(deltaPath))
-							.findFirst()
-							.isPresent();
-
-						if (needsRefresh) {
-							// TODO: there *may* be a remaining issue here if a
-							// source-generation plugin gets triggered
-							// by the above invocation of the jar:jar goal.
-							// This could cause Eclipse to think that the Java
-							// sources are dirty and queue the project
-							// for rebuilding, thus entering an infinite loop.
-							// One solution would be to find the output artifact
-							// jar and refresh ONLY that. However we
-							// have not been able to create the condition we
-							// are worried about so we are deferring any extra
-							// work on this until it's shown to be a
-							// real problem.
-							targetDir.refreshLocal(IResource.DEPTH_INFINITE, progress.newChild(1));
+						try {
+							SubMonitor progress = SubMonitor.convert(monitor, 3);
+							execJarMojo(projectFacade, progress.newChild(1, SubMonitor.SUPPRESS_NONE), isTest);
```

Comment:
I suggest using `split()` rather than `newChild()`. The main functional difference is that `split()` will efficiently do cancelation checks for you whereas `newChild()` will not do any cancelation checks. The only reason to use `newChild()` is if you need compatibility for Eclipse < 4.6 (which we don't).

There are a few other instances of `newChild()` here too. But this is an old issue so it might make sense to fix these in a new, dedicated PR.

Reply:
If there are several places that this needs tidying up then perhaps a separate bug should be raised and fixed? I would be keen to get rid of this build loop ASAP!

Reply:
@kriegfrj Makes sense, but I would agree with Tim, to do this in a separate PR. I will take care of this specific instance after this PR is through.

Reply:
Same to me. This bug makes Eclipse with bnd hard to use in our , slightly bigger project.